aten/src/ATen/native/Normalization.cpp

```diff
@@ -44,29 +43,23 @@ Tensor batch_norm(
   }
 
   bool use_cudnn = false;
-#if AT_CUDNN_ENABLED()
   use_cudnn = (input.type().is_cuda()
                && (input.type().scalarType() != at::kHalf
                  || weight.type().scalarType() == at::kFloat)
                && weight.defined() && bias.defined()
                && ((running_mean.defined() && running_var.defined())
                  || (!running_mean.defined() && !running_var.defined() && training))
                && input.size(0) <= 131070
-               && cudnn_enabled && CUDNN_VERSION >= 5110L);
-#else
-  // avoid unused parameter warning
-  (void)use_cudnn;
-  (void)cudnn_enabled;
-#endif
-
-#if AT_CUDNN_ENABLED()
-  if (use_cudnn && eps >= CUDNN_BN_MIN_EPSILON) {
+               && detail::getCUDAHooks().compiledWithCuDNN()
+               && cudnn_enabled && detail::getCUDAHooks().versionCuDNN() >= 5110L);
+
+  if (use_cudnn && eps >= detail::getCUDAHooks().batchnormMinEpsilonCuDNN()) {
     return std::get<0>(at::cudnn_batch_norm(
```

Comment:
Actually how does that work? Will dynamically loading `ATen_cuda` magically change this symbol? Or do we depend on lazy binding to make sure that this path will simply never be taken if `ATen_cuda` is not loaded, and won't cause missing symbol errors because of that?

Reply:
Not quite.

First, batchnormMinEpsilonCuDNN is a virtual function call defined in libATen_cpu.so, so we never have to worry about symbol resolution; the symbol is always defined, it's simply a matter of what object/vtable we are actually making the virtual function call on.  So how do we get our hands on the object?  We're using the Caffe2 registry, but what this effectively is, is a global variable which gets modified by the static initializer that runs when we dlopen('libATen_cuda.so')

OK, but as you say, can we just 'magically change the symbol'?  Here things can get a bit complicated.  This patch makes a simplifying assumption: once you make a call to ANY function that would require CUDA initialization, and it turns out that 'libATen_cuda.so' has NOT been loaded at this point, then we (1) give an error and (2) *permanently disable CUDA for the process.*

This is not intrinsic to the design, but it's easy to get the locking correct in this scheme.  One elaboration on this scheme, not for this PR, but Soumith/Yangqing/I agreed we should do, is have lazy CUDA initialization attempt to dlopen('libATen_cuda.so') *itself*, so that if you really do have CUDA it won't really be possible to get this wrong.